### DISCUSSION TITLE
WebP does not embed existing profile

### DISCUSSION THREAD

**Discussion by sethify** [libvips/libvips#2131] [Q&A]
I have a JPEG image with an `Display P3` profile (from IOS).

I have a conversion pipeline (in `C`) that performs transforms if required and writes out a new file in a specified format.

I have an input image (PNG) with an embedded `Display P3` profile.
My test suite shows outputs in PNG, GIF and JPEG formats retain the `Display P3` profile, however the WEBP files produced do not.  

I see here in the source you can provide an optional pointer to a profile:
https://github.com/libvips/libvips/blob/master/libvips/foreign/webpsave.c#L106

Is this something I must pass? If so, how do I get the pointer to the profile in the source image (using `C`).

I'm using LibVips version 8.10.5 installed with brew.

Here is the source image:
![gibson](https://user-images.githubusercontent.com/50456055/110149772-6f4b8800-7dac-11eb-8df1-ef8fc5b61555.png)

**Comment by jcupitt** [comment]
Hello @sethify,

This should work. I see:

```
$ vipsthumbnail guitar.png -o x.webp
$ vipsheader -f icc-profile-data guitar.png 
AAACJGFwcGwEAAAAbW50clJHQiBYWVogB+EABwAHAA0AFgAgYWNzcEFQUEwAAAAAQVBQTAAAAAAAAAAAAAAAAAAAAAAA
...
$ vipsheader -f icc-profile-data x.webp
AAACJGFwcGwEAAAAbW50clJHQiBYWVogB+EABwAHAA0AFgAgYWNzcEFQUEwAAAAAQVBQTAAAAAAAAAAAAAAAAAAAAAAA
...
```

Perhaps your libvips has not been built with webp metadata support? The webp library comes in two parts: libwebp and libwepdemux, with the demux part handling file format metadata (this was a terrible error on the part of the webp devs and the cause of a lot of bugs and confusion).

**Reply by sethify** [reply]
Thank you. It looks like I might have to start building the OSX version to ensure all the features are included. Is there a definitive build guide anywhere that explains the tradeoffs with the various libraries?

**Reply by jcupitt** [reply]
I had a quick look at homebrew's `vips.rb` and it looked like it ought to work. I should be able to get on a mac later today, I'll try and do a proper check.

The libvips README lists the various optional dependencies and has a few notes, but I'm sure it could be improved:

https://github.com/libvips/libvips/blob/master/README.md#dependencies

**Reply by sethify** [reply]
I've given that a good read. Two things I'm having trouble with are `gif` support in my custom build and `webp + profile`. I added the following to my build script and I'm still not getting that embedded profile (the profile embeds in other formats). I'm still digging.

```
# build webp
ENV WEBP_VERSION=1.2.0
ENV WEBP_URL=https://storage.googleapis.com/downloads.webmproject.org/releases/webp
      
RUN wget ${WEBP_URL}/libwebp-${WEBP_VERSION}.tar.gz \
  && tar xzf libwebp-${WEBP_VERSION}.tar.gz \
  && cd libwebp-${WEBP_VERSION} \
  && ./autogen.sh \
  && ./configure --enable-libwebpmux --enable-libwebpdemux --enable-libwebpdecoder \
  && make \
  && make install
```

**Reply by sethify** [reply]
This was the output during build if this helps:

```
WebP Configuration Summary
--------------------------

Shared libraries: yes
Static libraries: yes
Threading support: yes
libwebp: yes
libwebpdecoder: yes
libwebpdemux: yes
libwebpmux: yes
libwebpextras: no

Tools:
cwebp : yes
  Input format support
  ====================
  JPEG : yes
  PNG  : yes
  TIFF : yes
  WIC  : no
dwebp : yes
  Output format support
  =====================
  PNG  : yes
  WIC  : no
GIF support : yes
anim_diff   : yes
gif2webp    : yes
img2webp    : yes
webpmux     : yes
vwebp       : no
webpinfo    : yes
SDL support : no
vwebp_sdl   : no
```

**Reply by sethify** [reply]
I have confirmed that `/usr/lib` contains `libwebpdemux.so` and `libwebpdemux.so.2` as well as `libwebpmux.so` and `libwebpmux.so.3`. Maybe I'm missing a config flag on libvips?

**Reply by jcupitt** [reply]
I tried and it all seems to work for me with regular `brew install vips`.

I see:

```
22:00 $ which vips
/usr/local/bin/vips
✔ ~/pics 
22:00 $ vips --vips-config
native win32: no
native OS X: yes
open files in binary mode: no
enable debug: no
enable deprecated library components: yes
enable docs with gtkdoc: no
gobject introspection: yes
enable radiance support: yes
enable analyze support: yes
enable PPM support: yes
use fftw3 for FFT: yes
Magick package: MagickCore
Magick API version: magick7
load with libMagick: yes
save with libMagick: yes
accelerate loops with orc: yes
ICC profile support with lcms: yes (lcms2)
file import with niftiio: no
file import with libheif: yes
file import with OpenEXR: yes
file import with OpenSlide: yes
file import with matio: yes
PDF import with PDFium: no
PDF import with poppler-glib: yes
SVG import with librsvg-2.0: yes
zlib: yes
file import with cfitsio: yes
file import/export with libwebp: yes
text rendering with pangoft2: yes
file import/export with libspng: yes
file import/export with libpng: yes (pkg-config libpng >= 1.2.9)
support 8bpp PNG quantisation: yes
file import/export with libtiff: yes (pkg-config libtiff-4)
file import/export with giflib: yes (found by search)
file import/export with libjpeg: yes (pkg-config)
image pyramid export: yes
use libexif to load/save JPEG metadata: yes
✔ ~/pics 
22:00 $ vips copy guitar.png x.webp
✔ ~/pics 
22:01 $ vipsheader -a x.webp
x.webp: 500x667 uchar, 3 bands, srgb, webpload
width: 500
height: 667
bands: 3
format: uchar
coding: none
interpretation: srgb
xoffset: 0
yoffset: 0
xres: 1
yres: 1
filename: x.webp
vips-loader: webpload
icc-profile-data: 548 bytes of binary data
exif-data: 162 bytes of binary data
resolution-unit: in
exif-ifd0-Orientation: 1 (Top-left, Short, 1 components, 2 bytes)
exif-ifd0-XResolution: 72009/1000 (72.009, Rational, 1 components, 8 bytes)
exif-ifd0-YResolution: 72009/1000 (72.009, Rational, 1 components, 8 bytes)
exif-ifd0-ResolutionUnit: 2 (Inch, Short, 1 components, 2 bytes)
exif-ifd2-ExifVersion: Exif Version 2.1 (Exif Version 2.1, Undefined, 4 components, 4 bytes)
exif-ifd2-FlashpixVersion: FlashPix Version 1.0 (FlashPix Version 1.0, Undefined, 4 components, 4 bytes)
exif-ifd2-ColorSpace: 65535 (Uncalibrated, Short, 1 components, 2 bytes)
exif-ifd2-PixelXDimension: 500 (500, Long, 1 components, 4 bytes)
exif-ifd2-PixelYDimension: 667 (667, Long, 1 components, 4 bytes)
orientation: 1
xmp-data: 4689 bytes of binary data
✔ ~/pics 
```

So the generated webp file has an ICC profile attached. I didn't do anything special.

`otool` shows that my brew vips has all the webp libraries:

```
22:03 $ otool -L $(which vips) | grep -i webp
	/usr/local/opt/webp/lib/libwebpmux.3.dylib (compatibility version 4.0.0, current version 4.5.0)
	/usr/local/opt/webp/lib/libwebpdemux.2.dylib (compatibility version 3.0.0, current version 3.6.0)
	/usr/local/opt/webp/lib/libwebp.7.dylib (compatibility version 9.0.0, current version 9.0.0)
```

Is this an M1 mac, perhaps?

**Reply by sethify** [reply]
This is embarrassing. The image does have the profile data on my latest build. I was still convinced it doesn't because the Mac photo preview app recognizes colour profile on every other image type I've tested except webp. 

Thank you for digging into this with me. Very appreciated.

**Reply by jcupitt** [reply]
Oh no problem, I do much dumber things many times every day heh. Glad it's working.
